<html>
<head><title>Text2Gene: FAQ</title>
<meta name="title" content="Text2Gene: Frequently Asked Questions" />
<meta name="description" content="What benefits does Text2Gene have for clinical researchers in genetic testing?" />
<meta name="author" content="Text2Gene, LLC">
    <link rel="stylesheet" href="/text2gene.css" />
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
</head>
<body>

<div id="torso">
<h1>Frequently Asked Questions about Text2Gene</h1>

<h3><a href="#what">What is Text2Gene?</a></h3>

<p>Text2Gene is a service that finds clinically relevant evidence for human genetic variants in PubMed papers.</p>

<h3><a href="#howtouse">How do I use Text2Gene?</a></h3>

<p>To use the public version of Text2Gene, use the HGVS query form to enter in a genetic mutation string. The resultant page contains the original variant name entered, its gene name (if known), and a list of PubMed article entries mapped to their evidence sources, along with qualifying information, citation details, and publisher linkouts.</p>

<h3><a href="#why">Why do I need Text2Gene?</a></h3>

<p>Text2Gene rapidly finds evidence for rare human genetic variants (MAF &lt; .01) within published PubMed literature.</p>

<p>As clinical interpretation often hinges upon the quality and availability of evidence, finding this evidence has become the primary bottlneck for the speed and cost of genetic testing, even as the cost of genome sequencing declines.</p>

<p>Text2Gene performs the manual literature gathering that clinical scientists perform while finding evidence for a variant, while also broadening the reach using advanced text mining techniques.</p>

<p>The result: more clinically relevant annotations than you could find by manual searching alone, allowing you to <strong>save hundreds of hours of research time.</strong></p>


<h3><a href="#work">How does Text2Gene work?</a></h3>

<p>Text2Gene is an entirely automated, highly specialized search engine for human genetic variants (HGVS).</p>

<p>During the search, Text2Gene computes the logical lexical variations of a given variant, and uses these variant "synonyms" to find high-confidence mentions of the variant in PubMed articles. These articles are sourced from a combination of human-curated evidence databases (like ClinVar) and fulltext and/or abstract article indexes generated from clinical text processing.</p>

<h3><a href="#benefits">What benefits does Text2Gene provide over similar tools?</a></h3>

<p>Presently, no tool performs as comprehensive a search, finding as much relevant research, as Text2Gene. This is because Text2Gene makes use of many separate sources of hand-curated variant-evidence annotations alongside state-of-the-art clinical natural language processing tools.</p>

<p>Many tools rely heavily on experted-curated knowledge, but none of these sources individually can provide a set of evidence representing the best of what human curated annotations have to offer. By "casting a wide net" across reliable clinical databases Text2Gene catches many more evidence sources than searching any one alone.</p>

<p>Text2Gene leverages state-of-the-art clinical NLP (natural language processing) techniques to locate variants within novel evidence sources, many of which have not yet been accessioned into the human curated evidence databases. 

<p>The benefits of combining these two powerful approaches comprise the ultimate goal of Text2Gene: saving time researching clinical variants, and finding more highly relevant evidence towards clinical interpretation.</p>

<h3><strong><a href="#nocitation">Why does Text2Gene show no evidence for my variant?</a></h3>

<p>The most common reason why Text2Gene cannot find evidence for a given variant is that the variant has not yet been well characterized in published literature.</p>


<h3><a href="#whichindexes">Which indexes does Text2Gene use?</a></h3>

<p>The public version of Text2Gene indexes variant mentions across the following sources:</p>
<ul>
<li>ClinVar</li>
<li>NCBI</li>
<li>PubMed Abstracts (via tmVar)</li>
<li>Google custom search engine</li>
</ul>

<p>Proprietary sources (such as HGMD and COSMIC) can be integrated on a per-institution basis (with proper licensing agreements). Please contact <a href="mailto:info@text2gene.com">info@text2gene.com</a>.

<hr />
<p><a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC4544753/">According to the ACMG</a>, "Increasingly, clinical molecular laboratories are detecting novel sequence variants in the course of testing patient specimens for a rapidly increasing number of genes associated with genetic disorders."  </p>

</div>
</body>
</html>
